### PR TITLE
Adds `grpc` field to `liveness_probe` and `startup_probe` to `google_cloud_run_v2_service` resource

### PR DESCRIPTION
**Pull request by mmizutani**
This PR adds gRPC probes options for [Cloud Run v2 Service resources](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloud_run_v2_service).

The `grpc` fields in `startup_probe` and `liveness_probe` in the `google_cloud_run_service` (v1) resource were added by #6781.
The support for Cloud Run v2 Services was introduced by #6850 and released in [`terraform-google-provider` 4.46.0](https://github.com/hashicorp/terraform-provider-google/blob/main/CHANGELOG.md#4460-december-12-2022), but the gRPC options for liveness probe and startup probe for the v2 services are missing.

The Go SDK v0.102.0, which this project currently uses, [does not yet support the gRPC probe options for Cloud Run v2 Services](https://pkg.go.dev/google.golang.org/api@v0.102.0/run/v2#GoogleCloudRunV2Probe). The SDK needs to be upgraded to at least [v0.104.0](https://pkg.go.dev/google.golang.org/api@v0.104.0/run/v2#GoogleCloudRunV2Probe).
So this PR waits for #6962 to be merged.

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
added `template.0.containers0.liveness_probe.grpc`, `template.0.containers0.startup_probe.grpc` fields to `google_cloud_run_v2_service` resource
```

### CODE REVIEW

**Comment by ScottSuarez** [comment]
Hello I notice this PR is in draft. Please let me know if anything is needed!

**Comment by mmizutani** [comment]
Thanks. I am waiting for #6962 to be merged.

File: GNUmakefile
Comment (mmizutani) [inline GNUmakefile:74]: Adding this make target was necessary to run the non-e2e tests locally as per the developer guide:

> Run unit tests
> Unit tests (that is, tests that do not interact with the GCP API) are very fast and you can generally run them all if you have changed any of them:
> 
> make test

https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md#run-unit-tests

File: mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go
Comment (mmizutani) [inline mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go:454]: Google Cloud does not provide any gRPC variant of official container images for Cloud Run Services. `gcr.io/cloudrun/hello` is only for REST API and Cloud Event processing servers.

To make this E2E tests more realistic, it would be desirable to use a container image of a real gRPC server that implements the gRPC health check service.
So I [created](https://github.com/mmizutani/cloud-run-hello-grpc) such a custom docker image, but it would be more desirable to use more "official" container images hosted by Google Cloud or Hashicorp.

**Comment by ScottSuarez** [comment]
Hi mmi, I notice this pull request is in draft.. please mark as ready for review when a pass is needed

File: mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go
Comment (mmizutani) [inline mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go:492]: This gRPC server container image is hosted in my personal temporary repository.
As explained in [`mmv1/third_party/terraform/utils/test-fixtures/cloudrun/hello-cloud-run-grpc/README.md`](https://github.com/mmizutani/magic-modules/blob/cloud-run-v2-grpc-probe/mmv1/third_party/terraform/utils/test-fixtures/cloudrun/hello-cloud-run-grpc/README.md), this custom image is needed for testing the gRPC probe of the `google_cloud_run_v2_service` resource, but no official image of ready-to-use, prebuilt gRPC server container image for Cloud Run is provided by Google Cloud or gRPC schosystem developers.

It would be nice if we could host this container image in the same Google Cloud project as `gcr.io/graphite-docker-images/magic-modules-dev`. I am thinking of adding a GitHub Actions workflow like [`magic-modules/.ci/gcb-downstream-builder.yml`](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/gcb-downstream-builder.yml) that runs only when files under `mmv1/third_party/terraform/utils/test-fixtures/cloudrun/hello-cloud-run-grpc/` change in the main branch.

Otherwise, we need to build this container image during each run of the `google_cloud_run_v2_service` service, but we would end up with spending majority of the time for tests in testing this docker build workflow rather than in testing Cloud Run services.
Theoretically it could be possible to automate this in `make test`, but this requires adding the Go Docker client library and Docker/Podman runtime to CI dependencies.

File: mmv1/products/cloudrunv2/api.yaml
Comment (ScottSuarez) [inline mmv1/products/cloudrunv2/api.yaml:1028]: Why allow sending empty object?

File: mmv1/products/cloudrunv2/api.yaml
Comment (ScottSuarez) [inline mmv1/products/cloudrunv2/api.yaml:1032]: commented out code... can be omitted.

File: mmv1/products/cloudrunv2/api.yaml
Comment (ScottSuarez) [inline mmv1/products/cloudrunv2/api.yaml:1043]: Can one of these values be required.. having empty blocks in terraform code to perform behavior is confusing for customers.

File: GNUmakefile
Comment (ScottSuarez) [reply GNUmakefile:76]: This documentation is specific to the downstream providers.. Lets omit this

File: mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go
Comment (ScottSuarez) [reply mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go:492]: We definitely will not want to host this in a contributor project. Is there a publicly accessible image we can rely on? I don't like the idea of managing an application image for a single test.

Comment (mmizutani) [reply mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go]: >  Is there a publicly accessible image we can rely on?

Unfortunately, no. 
I did search for publicly accessible container images of a gRPC server with the standard health check implementation, but I was not able to find any. Neither Google Cloud nor the gRPC ecosystem team provide such a convenient free image.

Comment (mmizutani) [reply mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go]: Only the grpc startup probe, not the grpc liveness probe, requires an actually working gRPC server.
So I concluded that in the meantime I will postpone adding support for the grpc block in the startup probe in this PR.

Comment (mmizutani) [reply mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go]: >  Is there a publicly accessible image we can rely on?

Unfortunately, no. 
I did search for publicly accessible container images of a gRPC server with the standard health check implementation, but I was not able to find any. Neither Google Cloud nor the gRPC ecosystem team provide such a convenient free image.

Only the grpc startup probe, not the grpc liveness probe, requires an actually working gRPC server.
So I concluded that in the meantime I will postpone adding support for the grpc block in the startup probe in this PR.

File: mmv1/products/cloudrunv2/api.yaml
Comment (mmizutani) [reply mmv1/products/cloudrunv2/api.yaml:1045]: If the policy of the Terraform Google Provider is not just simply wrap the Google Cloud SDK and replicate its original behavior precisely as possible but reduce inconvenience and awkward behaviors by wrapping it with easier interface, then I would go mark the `port` property as required.
However, in that case, it would be better to also change the following properties from optional to required so that users of this resource will not get confused:

- `template.containers.startupProbe.httpGet.path`
- `template.containers.startupProbe.tcpSocket.port`
- `template.containers.livenessProbe.httpGet.path`
- `template.containers.livenessProbe.tcpSocket.port`

Comment (ScottSuarez) [reply mmv1/products/cloudrunv2/api.yaml:1045]: Yep! We like to reduce inconvenience by making enforcing terraform blocks to not be empty by design..

If we can do this and the experience isn't negatively affected. Lets go for it. Although if it will be a net negative impact or there are reasons to the contrary for these particular fileds then please let me know!

File: mmv1/products/cloudrunv2/api.yaml
Comment (ScottSuarez) [reply mmv1/products/cloudrunv2/api.yaml:1034]: Sure! So this brings up the topic of `exactly_one_of` and what these commented out blocks were originally trying to ensure. 

`exactly_one_of` allows terraform to enforce that at least one of the given values are set. Unfortunately `exactly_one_of` and its sisters (`at_least_one_of` and `conflicts_with`) don't work under nested, itterable fields.

https://github.com/hashicorp/terraform-plugin-sdk/issues/470

The intention of these comments is to leave room for this behavior to be enabled at some future point in time although I don't see this issue being resolve in the near to moderate future. Thus I would be in favor of omitting these comments in future additions. It reduces code clarity with little value in my opinion.

File: mmv1/products/cloudrunv2/api.yaml
Comment (ScottSuarez) [reply mmv1/products/cloudrunv2/api.yaml:1030]: We try to avoid allow empty object since we are enabling features with empty blocks. From a terraform usage PoV it can be pretty confusing! Let me see what our teams position is on this... My immediate thought/preference is to make port required.

Usually apis have a separate field for feature enablement.
```hcl
resource "x" "y" {
  grpc {
    enabled = true
   }
}
```

Which allows the enablement of the feature to be more explicit. 

In my opinion an empty grpc block for feature enablement is slightly confusing to me with my background. 
```hcl
resource "x" "y" {
  grpc {
   }
}
```

let me see what my team says.

Comment (ScottSuarez) [reply mmv1/products/cloudrunv2/api.yaml:1030]: response
```
Empty blocks should be avoided when possible. They are not well represented in Terraform.

`grpc{}` in understood by Terraform as []interface{}{nil}, not []interface{}{map[string]interface{}{}}
```

File: mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go
Comment (ScottSuarez) [reply mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go]: Maybe we can use something like 
https://github.com/GoogleCloudPlatform/golang-samples/tree/f04b64bb5e595bcf7d3726a7eb7ed0d0557ed90a/run/grpc-ping#deploying-to-cloud-run 

Although this might have a prerequisite of building the image. The other alternative would be to use an image that doesn't work correctly and expect the deployment to fail. We would still be exercising the APIs acceptance of these fields which would be sufficient for our coverage.

File: mmv1/products/cloudrunv2/api.yaml
Comment (mmizutani) [reply mmv1/products/cloudrunv2/api.yaml]: Given the lack of overall guidance (https://github.com/hashicorp/terraform-provider-google/issues/12648) on whether or not we should leave `exactly_one_of` and so on under repeatable nested fields as a comment, I will follow your advice and omit these comment lines at least here. 
There are a bit too many `exactly_one_of` comments in YAML files.

File: mmv1/products/cloudrunv2/api.yaml
Comment (mmizutani) [reply mmv1/products/cloudrunv2/api.yaml]: It is better to honor the convention of the Terraform language syntax and prevent surprise to the provider users, so I will omit `send_empty_values: true` and `allow_empty_object: true` and mark `grpc.port` as required. 
In most cases, explicitly specifying a health check port number will not cause any notable inconvenience to Terraformers, and making infrastructure configurations explicit as code rather than implicit set will be usually preferred.

File: mmv1/products/cloudrunv2/api.yaml
Comment (yanweiguo) [inline mmv1/products/cloudrunv2/api.yaml:1027]: Could you remove `If not specified, defaults to 8080.`? It's not accurate. The cloud run API documents will be updated as well.

Also appreciated if you can update the description for other probe ports.

File: mmv1/products/cloudrunv2/api.yaml
Comment (yanweiguo) [reply mmv1/products/cloudrunv2/api.yaml:1033]: It allows empty blocks for tcpSocket and httpGet.

In cloudrun v1 Terraform, it also allows empty blocks: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/cloudrun/api.yaml#L711

Shall we keep consistency?

File: mmv1/products/cloudrunv2/api.yaml
Comment (mmizutani) [reply mmv1/products/cloudrunv2/api.yaml:1027]: @yanweiguo 

> Could you remove If not specified, defaults to 8080.? It's not accurate. The cloud run API documents will be updated as well.

Noted. Then, what would be the correct behavior of the Cloud Run Admin API? How about the revised description below?
```suggestion
                            Port number to access on the container. Number must be in the range 1 to 65535. If not specified, defaults to the port number that the gRPC service is actually listening on.
```


> Also appreciated if you can update the description for other probe ports.

Although understandable, that is beyond the scope of this pull request. Could you address the rest part in your separate pull request?

Comment (yanweiguo) [reply mmv1/products/cloudrunv2/api.yaml:1027]: It defaults to the container port, which is the value of `container.ports[0].containerPort`.

It's not the port gRPC service is actually listening on since we don't know the actual behavior inside the container.

File: mmv1/products/cloudrunv2/api.yaml
Comment (mmizutani) [reply mmv1/products/cloudrunv2/api.yaml:1033]: It seems that there are differing preferences among the maintainers regarding to what extent Terraform Google Provider should mirror the actual behavior of the Google Cloud APIs - Terraform should precisely replicate every bit of the behaviors of the Google Cloud APIs and keep consistency with other parts of this repository, or it should work more as a user-friendly wrapper of the APIs consistently with the existing convention of Terraform providers.

Would it be possible to move this discussion topic to an issue ticket and make this pull request settled in a temporary middle ground?
It is beyond the scope of this pull request, and there will be lots of code in this repository that is relevant to this coding policy.
By moving the discussion to an issue ticket, we will be able to collect opinions from other developers as well.

Comment (shuyama1) [reply mmv1/products/cloudrunv2/api.yaml:1033]: We generally would avoid allowing users sending empty blocks as it always causes more problems and some of them are hard to fix. However, if I remember correctly, based on some discussion in other PRs or offline, sending empty blocks like `tcp_socket{}` and `http_get{}` is considered different from not sending the field at all in this case. That's probably why we allow empty blocked in the first place for v1 resource, and v2 does not change the behavior (please correct me if that's not the case), so I just keep the consistency when implementing it.

@ScottSuarez do you know if we should allow users to send empty blocks if sending them is different than unset of the field?

**Comment by ScottSuarez** [comment]
Could you pull the latest changes in from main? 

Also! feel free to ping when when you want another review or you want some help clarifying things.. Happy to talk :)

File: mmv1/products/cloudrunv2/api.yaml
Comment (yanweiguo) [reply mmv1/products/cloudrunv2/api.yaml:1033]: Yeah, one and only one of `tcp_socket`, `http_get` and ` `startupProbe` must be set in `startup_probe`. One and only one of `http_get` and ` `startupProbe` must be set in `liveness_probe`. Otherwise Cloud Run API will throw an error.

**Comment by ScottSuarez** [comment]
Looks like still an outstanding test failure

```
=== CONT  TestAccCloudRunV2Service_cloudrunv2ServiceGRPCProbesUpdate
    provider_test.go:315: Step 1/8 error: Error running apply: exit status 1
        
        Error: Error creating Service: googleapi: Error 400: Violation in CreateServiceRequest.service.template.containers[0].liveness_probe.probe_type: required oneof field 'probe_type' must have one initialized field
        Details:
        [
          {
            "@type": "type.googleapis.com/google.rpc.BadRequest",
            "fieldViolations": [
              {
                "description": "required oneof field 'probe_type' must have one initialized field",
                "field": "Violation in CreateServiceRequest.service.template.containers[0].liveness_probe.probe_type"
              }
            ]
          }
        ]
        
          with google_cloud_run_v2_service.default,
          on terraform_plugin_test.tf line 3, in resource "google_cloud_run_v2_service" "default":
           3: resource "google_cloud_run_v2_service" "default" {
        
--- FAIL: TestAccCloudRunV2Service_cloudrunv2ServiceGRPCProbesUpdate (2.04s)

```

**Comment by mmizutani** [comment]
Let me summarize the current situation of this pull request review.
This pull request has been in an intricate deadlock due to the following conflicting suggestions based on different perspectives from different reviewers regarding the optionality of the fields of the `service.template.containers[0].liveness_probe.grpc` and `service.template.containers[0].startup_probe.grpc` blocks.

- @ScottSuarez [advised](https://github.com/GoogleCloudPlatform/magic-modules/pull/6987#discussion_r1072740090) that at least one of the fields of the `grpc` blocks should be marked as required since enabling features with empty blocks is unintuitive and confusing from the perspective of the general Terraform convention.
- @yanweiguo [suggested](https://github.com/GoogleCloudPlatform/magic-modules/pull/6987#discussion_r1098058477) that the grpc blocks should be allowed to be empty to be in line with the v1 run resources as well as the `http_get` and `tcp_socket` blocks of the v1/v2 run resources.
- @shuyama1 [implied](https://github.com/GoogleCloudPlatform/magic-modules/pull/6987#discussion_r1100566485) that the v1 run resource and v2 run resource `http_get` and `tcp_socket` blocks intentionally allowed to be empty in Terraform syntax in order to send empty fields to the Google Run Admin API since requests missing the fields and requests with empty fields are interpreted differently by the API.

I have confirmed that @shuyama1's point is likely true. When I allow empty `grpc` blocks in Terraform and the following v2 run resource with an empty grpc block, I observed the following POST request with an JSON body containing an empty grpc block field rather than a null grpc field, and the API indeed interpreted the empty block fields and the null fields differently. 
```hcl
resource "google_cloud_run_v2_service" "default" {
  name     ="%{service_name}"
  location = "us-central1"

  template {
    containers {
      image = "us-docker.pkg.dev/cloudrun/container/hello"
      ports {
        container_port = 8080
      }
      liveness_probe {
        grpc {}
      }
    }
  }
}
```
```
---[ REQUEST ]---------------------------------------
POST /v2/projects/PROJECT_ID/locations/us-central1/services?alt=json&serviceId=tf-test-cloudrun-serviceuut6x9qm0w HTTP/1.1
Host: run.googleapis.com
User-Agent: Terraform/1.3.6 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/acc
Content-Length: 237
Content-Type: application/json
Accept-Encoding: gzip

{
 "template": {
  "containers": [
   {
    "image": "us-docker.pkg.dev/cloudrun/container/hello",
    "livenessProbe": {
     "failureThreshold": 3,
     "grpc": {}, /// {} not null
     "httpGet": null,
     "periodSeconds": 10,
     "tcpSocket": null,
     "timeoutSeconds": 1
    },
    "ports": [
     {
      "containerPort": 8080
     }
    ]
   }
  ]
 }
}
```
As long as sending a JSON request body with empty, non-null fields requires writing empty blocks in Terraform code in google-terraform-provider, it seems that we need to allow `template.containers[0].liveness_probe.grpc` and `service.template.containers[0].startup_probe.grpc` blocks to be empty just like `http_get` and `tcp_socket` as they are today.
Given such tight coupling of the Terraform block syntax and JSON API request body, to avoid confusing situations where users have to enable features using empty Terraform blocks, we would need to either refactor the syntax of `google_cloud_run_v2_service` to allow enabling features without writing empty blocks or simply mark at least one of the fields of the blocks as required. The former of course introduces breaking changes and would be unacceptable. The latter is much easier to implement but causes inconsistency with the existing siblings that allow setting empty blocks (v1 run resource and v2 run resource's `http_get` and `tcp_socket` blocks), and changing the existing `http_get` and `tcp_socket` blocks from allowing empty blocks to disallowing empty blocks also breaks backward compatibility. 
Introducing a new field `enabled = true|false` within the `grpc` block is awkward here since `grpc` is one of the `exactly_one_of` blocks. 

To move things forward, I would suggest that we allow `service.template.containers[0].liveness_probe.grpc` and `service.template.containers[0].startup_probe.grpc` to be set empty just like similar, other parts of this repository and move the continuation of the discussion on the appropriateness and alternatives of the currently allowed empty blocks syntax to a standalone repository issue ticket as this issue pertains to lots of other parts of this repository as well.

Could you discuss among you three reviewers and reconcile opinions as to how to proceed?

**Comment by ScottSuarez** [comment]
Hi @mmizutani, let go with your suggestion. Unfortunately there are scenarios, as you mentioned, where we need to utilize empty blocks. It's not an ideal design pattern from my point of view but it's not a blocker. You can go ahead and implement it as you think best and we can work to get that in. I'll follow up with any clarifying questions.

File: mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go
Comment (mmizutani) [reply mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go]: > The other alternative would be to use an image that doesn't work correctly and expect the deployment to fail. We would still be exercising the APIs acceptance of these fields which would be sufficient for our coverage.

I have followed this course of your advice and added this type of tests: 
https://github.com/mmizutani/magic-modules/blob/68ddd69ac609fc9c113af93cefca600c43d55685/mmv1/third_party/terraform/tests/resource_cloud_run_v2_service_test.go#L304-L328

**Comment by mmizutani** [comment]
@ScottSuarez 
Acceptance tests for the google_cloud_run_v2_service resource are passing in my local environment. Could you trigger the `generate-diffs` and `run-rake-tests` CI workflows again?

**Review by ScottSuarez — APPROVED** [review]
Thanks for your work and patience @mmizutani :)

**Pull request merged**
